garak/generators/azure.py

```diff
@@ -0,0 +1,116 @@
+"""Azure OpenAI generator
+
+Supports chat + chatcompletion models. Put your API key in the AZURE_API_KEY environment variable, 
+the azure openai endpoint in the AZURE_ENDPOINT environment variable and the azure openai deployment name 
+in AZURE_DEPLOYMENT_NAME environment variable.
+
+Put the model name in either the --model_name command line parameter, or
+pass it as an argument to the Generator constructor.
+"""
+
+import os
+import openai
+
+from garak.generators.openai import OpenAICompatible, chat_models, completion_models, context_lengths
+
+# lists derived from https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models
+# some azure openai model names should be mapped to openai names
+openai_model_mapping = {
+    "gpt-4": "gpt-4-turbo-2024-04-09", 
+    "gpt-35-turbo": "gpt-3.5-turbo-0125", 
+    "gpt-35-turbo-16k": "gpt-3.5-turbo-16k", 
+    "gpt-35-turbo-instruct": "gpt-3.5-turbo-instruct"
+}
+
+class AzureOpenAIGenerator(OpenAICompatible):
+    """Wrapper for Azure Open AI. Expects AZURE_API_KEY, AZURE_ENDPOINT and AZURE_DEPLOYMENT_NAME environment variables.
+
+    Uses the [OpenAI-compatible API](https://learn.microsoft.com/en-us/azure/ai-services/openai/api-version-deprecation)
+    via direct HTTP request.
+
+    To get started with this generator:
+    #. Visit [https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models) and find the LLM you'd like to use.
+    #. [Deploy a model](https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/create-resource?pivots=web-portal#deploy-a-model) and copy paste the model and deployment names
+    #. On the Azure portal page for the Azure OpenAI you want to use click on "Resource Management -> Keys and Endpoint" and copy paste the API Key and endpoint. 
+    #. In your console, Set the ``AZURE_API_KEY``, ``AZURE_ENDPOINT`` and ``AZURE_DEPLOYMENT_NAME`` variables.
+    #. Run garak, setting ``--model_type`` to ``azure`` and ``--model_name`` to [the name of the model](https://learn.microsoft.com/en-us/azure/ai-services/openai/concepts/models)
+    - e.g. ``gpt-4o``.
+    """
+
+    ENV_VAR = "AZURE_API_KEY"
+    DEPLOYMENT_NAME_ENV_VAR = "AZURE_DEPLOYMENT_NAME"
+    ENDPOINT_ENV_VAR = "AZURE_ENDPOINT"
+
+    active = True
+    generator_family_name = "Azure"
+    api_key = None
+    deployment_name = None
+    endpoint = None
+    name_backup = ""
```

Comment:
Default instance values are set via `DEFAULT_PARAMS` due to `garak`'s [`Configurable`](https://github.com/leondz/garak/pull/711) pattern for plugins.

```suggestion
    DEFAULT_PARAMS = OpenAICompatible.DEFAULT_PARAMS | {
        deployment_name: None,
        endpoint: None,
        name_backup: None,
    }
```

This is intentional as class level variable are shared by all instances of the class and changes to the values by an instance could leak data to other instances.

Note that `api_key` is removed as `garak` plugins with an `ENV_VAR` account for this instance value.